### ISSUE TITLE
[Workflow] Be able to know why exactly transitions can't be made.

### ISSUE THREAD

**Issue by d-ph** [symfony/symfony#24501]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.3.10

Hello,

Would it be possible to add a way to know exactly why a transition isn't possible, please?

Motivation: I'd like to show to my Users why exactly a transition couldn't happen. For this I need to know why it can't happen. Currently, the `Workflow::can(): bool` returns `true` or `false`. Ideally, it would instead return a collection of "TransitionViolation[]" (similar to how `ValidatorInterface::validate(): ConstraintViolationListInterface` returns a collection of `ConstraintViolation`). This way I can block a transition with a `TransitionViolation` and then retrieve those violations in the code, that tries to transition objects' statuses, and eventually show a descriptive reason to my Users, what they need to fix in order to move forward.

In other words, in a `Guard` the following would be possible:
```php
public function onTransition(GuardEvent $event)
{
    if (is_monday()) {
        /*
         * addTransitionViolation() would internally call $event->setBlocked(true);
         */
        $event->addTransitionViolation(new TransitionViolation('Cannot transition on Monday.'));
    }
}
```

Problems:
2.1. Changing the return type of `Workflow::can()` is a BC break. Introducing a new method would be better. E.g. `Workflow::canTransitionReturnWhyNot(): TransitionViolation[]`. Picking up a nice name for the new method will be a challenge.
2.2. `TransitionViolation` might not be the best name under the Sun. I'm open for suggestions. E.g. `TransitionRefusal`, `GuardRefusal` or similar.

`TransitionViolation` should accept a message, message parameters (optional) and a machine error code (optional):
```php
class TransitionViolation 
{
    public function __construct(string $message, array $messageParameters = [], string $errorCode = 'default') {}
}

/*
 * Example usage
 */
$foo = new TransitionViolation(
    'Cannot transition, when the distance difference is less than 5 kilometres',
    ['minDistance' => 5],
    'com.symfony.www.min_distance'
);
```

Thank you.

**Comment by lyrixx** [comment]
This proposal looks good to me, and the proposed API is easy to use. I like it.
I guess the translator could be involved here. Like in the validator component.

And yes, the naming will be hard.

Anyway, I'm :+1: with the idea.

Do you want to contribute to this new feature?

**Comment by d-ph** [comment]
I'll take a look the weekend after this weekend.

I'll need to change how `Workflow::apply()` works. Long story short: I stopped calling 
```
if ($workflow->can($subject, $transitionName)) {
  $workflow->apply($subject, $transitionName);
}
```
in my code, because this runs the guards twice and I have some slightly not cheap checks in my guards, which I'd rather not run more times than needed during a request. Instead I call `Workflow::apply()` in a try..catch block, catching the Workflow's `LogicException`, like this:
```
try {
  $workflow->apply($subject, $transitionName);
} catch (LogicException $exception) {
  // return an http response saying "No".
}
```
The thing with `Workflow::apply()` is that it throws `LogicException` in two different cases, but that's fine in my code, because I know for a fact, that at that point of the execution, only one of those two cases may occur (i.e. transition can't be made as opposed to transition isn't defined).

After the new feature is added, I'd like to continue running `Workflow::apply()` in a try..catch, but also would like to know the collection of `TransitionViolation[]`. Are you ok if I create a new Workflow exception class (e.g. `LogicExceptionWithTransitionViolations`, I need to think of a better name for this), that inherits `LogicException`, so there's no BC break, and throw that instead?

There's still the problem of `Workflow::apply()` throwing `LogicException` in two different cases, because now I'd like to throw `LogicException` for the case, when the requested transition isn't defined, and throw `LogicExceptionWithTransitionViolations`, when the transition is defined, but can't be made. I'll change `Workflow::apply()` to do just that, by explicitly retrieving the requested transition(s) by name (as opposed to retrieving all enabled transition for `$subject`'s current state and then matching the name in a `foreach` loop) and then first checking whether I managed to retrieve it (is it defined?), and secondly checking, whether I can transition to it. There'll probably be more hurdles along the way, but I just wanted to know, whether you're ok for me to change the `Workflow::apply()` to achieve the described behaviour.

Thanks,
@d-ph
